dask/dataframe/tests/test_csv.py

```diff
@@ -0,0 +1,129 @@
+
+
+def test_bytes_read_csv_with_header():
+    b = files['2014-01-01.csv']
+    header, b = b.split(b'\n', 1)
+    df = bytes_read_csv(b, header, {})
+    assert list(df.columns) == ['name', 'amount', 'id']
+    assert len(df) == 3
+    assert df.id.sum() == 1 + 2 + 3
+
+
+def test_read_csv():
+    bytes = [files[k] for k in sorted(files)]
+    gzbytes = [gzip_compress(b) for b in bytes]
+    kwargs = {}
+    head = bytes_read_csv(files['2014-01-01.csv'], '', {})
+
+    for blocks, compression in [(bytes, None),
+                                (gzbytes, 'gzip')]:
+        blocks = [[b] for b in blocks]
+        kwargs = {'compression': compression}
+
+        df = read_csv_from_bytes(blocks, header, head, kwargs, collection=True)
+        assert isinstance(df, dd.DataFrame)
+        assert list(df.columns) == ['name', 'amount', 'id']
+
+        values = read_csv_from_bytes(blocks, header, head, kwargs,
+                                     collection=False)
+        assert isinstance(values, list)
+        assert len(values) == 3
+        assert all(hasattr(item, 'dask') for item in values)
+
+        f = df.amount.sum().compute()
+        result = df.amount.sum().compute()
+        assert result == (100 + 200 + 300 + 400 + 500 + 600)
+
+
+def test_kwargs():
+    blocks = [files[k] for k in sorted(files)]
+    blocks = [[b] for b in blocks]
+    kwargs = {'usecols': ['name', 'id']}
+    head = bytes_read_csv(files['2014-01-01.csv'], '', kwargs)
+
+    df = read_csv_from_bytes(blocks, header, head, kwargs, collection=True)
+    assert list(df.columns) == ['name', 'id']
+    result = df.compute()
+    assert (result.columns == df.columns).all()
+
+
+def test_blocked():
+    blocks = []
+    for k in sorted(files):
+        b = files[k]
+        lines = b.split(b'\n')
+        blocks.append([b'\n'.join(bs) for bs in partition_all(2, lines)])
+
+    df = read_csv_from_bytes(blocks, header, expected.head(), {})
+    eq(df.compute().reset_index(drop=True),
+       expected.reset_index(drop=True), check_dtype=False)
+
+    expected2 = expected[['name', 'id']]
+    df = read_csv_from_bytes(blocks, header, expected2.head(),
+                             {'usecols': ['name', 'id']})
+    eq(df.compute().reset_index(drop=True),
+       expected2.reset_index(drop=True), check_dtype=False)
+
+
+def test_read_csv_files():
+    with filetexts(files, mode='b'):
+        df = read_csv('2014-01-*.csv')
+        eq(df, expected, check_dtype=False)
+
+
+from dask.bytes.compression import compressors
+fmt_bs = [(None, None), (None, 10), ('gzip', None), ('bz2', None),
+          ('xz', None), ('xz', 10)]
+fmt_bs = [(fmt, bs) for fmt, bs in fmt_bs if fmt in compressors]
+
+@pytest.mark.parametrize('fmt,blocksize', fmt_bs)
+def test_read_csv(fmt, blocksize):
+    compress = compressors[fmt]
+    files2 = valmap(compress, files)
+    with filetexts(files2, mode='b'):
+        df = read_csv('2014-01-*.csv', compression=fmt)
+        eq(df, expected, check_dtype=False)
```

Comment:
@wabu can you try this implementation against your dataset and compare performance?  I would actually be pretty interested in the result of profiling.  I might do this in the following way.

```python
import dask
from dask.dataframe.csv import read_csv

df = read_csv(path, compression='xz')
```

I would do this in the IPython console and then use the `%prun` magic

```python
%prun df.some_column.sum().compute(get=dask.get)  # single threaded scheduler
```